Make build parallel safe

mattiaverga opened the pull request:
While packaging XEphem for inclusion in Fedora/EPEL I found that the current Makefile doesn't behave well when the buildsystem is using parallel compilation.

I'm not an expert here, but this change should make libastro compiled first, then other libraries and finally xephem executable.

mattiaverga committed:
Make build parallel safe

brandon-rhodes commented:
@mattiaverga — While it may be a few weeks before I can dig into your other PRs, this one is straightforward enough that I can go ahead and ask some quick questions:

1. Why are libraries like `libpng` omitted? Wouldn't that cause the build to fail on platforms that lack them? Or do they still get built and I'm just not seeing quite how?
2. Does this break the target `make libs`, for folks who have been typing that or calling it from other scripts?
3. Are the two libraries `libip liblilxml` singled out as special because they're the only two that depend on libastro?

mattiaverga commented:
> 
>     1. Why are libraries like `libpng` omitted? Wouldn't that cause the build to fail on platforms that lack them? Or do they still get built and I'm just not seeing quite how?
My mistake, I've committed the changes I made for Fedora, where we use system libraries, without taking other cases in consideration. I've now corrected by adding the third party libraries to the `LIBRARIES` array.
> 
>     2. Does this break the target `make libs`, for folks who have been typing that or calling it from other scripts?
Yes, because the `libs` target didn't exist anymore. I've added it again as a placeholder for building the `LIBRARIES` array.
> 
>     3. Are the two libraries `libip liblilxml` singled out as special because they're the only two that depend on libastro?
If I understood well how the make parallelization works, I'm saying to the build system that the `xephem` target depends on `$(LIBRARIES)` and that `LIBRARIES` depends on libastro. So, even using parallel build, libastro will be built first alone, then all libraries are built in parallel and finally the xephem executable target is run.

rmathar commented:
A Fedora build is in https://build.opensuse.org/package/show/home:rjmathar/xephem . One of the first steps in that compilation under rpm is to remove libXm, libjpegd, libpng and libz of the XEphem source and replace these by the OS libraries, see the file xephem.spec .

brandon-rhodes commented:
@mattiaverga — Thanks for your patience while I had a busy October and November—I was invited to give my first in-person talk since the pandemic, and so was writing the talk then traveling to the code::dive conference in Wrocław to deliver it.

So, reading over this diff again:

* It exposes to `make` that each library's compilation can proceed independently, instead of forcing them to build in series. Which is definitely an improvement.
* It tells `make` that building the final binary needs the libraries, whereas previously this dependency was not explicit. Also an improvement!
* It uses two features that I'm not sure XEphem has used before in its `Makefile`: the `$(MAKE)` variable, and the `--directory=` command line option.

That last item worries me—several XEphem enthusiasts run it on old 1990s workstations, and they are using a traditional UNIX `make` instead of GNU Make, and so the double-dash options at least are not likely to be present.

I suppose what I should really do, rather than send you off to perform tests on an old platform or else to consult 1990s UNIX docs, is to set up a CI test step here on GitHub that would automatically verify that compilation can proceed on an example of an old Unix. I wonder where I might find an emulator? I'll look for one when I have free time over the next few days and will report back!

rmathar commented:
There is a famous quote from Max Planck about progress in general: "Eine neue wissenschaftliche  Wahrheit pflegt sich nicht in der Weise durchzusetzen, dass ihre Gegner überzeugt werden und sich als belehrt erklären, sondern vielmehr dadurch, dass ihre Gegner allmählich aussterben und dass die heranwachsende Generation von vornherein mit der Wahrheit vertraut gemacht ist." , translated in https://en.wikipedia.org/wiki/Planck%27s_principle
